concent_api/utils/tests/test_api_view.py

```diff
@@ -115,6 +123,120 @@ def dummy_view(request, message):                                       # pylint
         self.assertEqual(response['content-type'], "application/json")
         self.assertEqual(response.status_code, 415)                             # pylint: disable=no-member
 
+    def test_request_with_not_allowed_http_method_should_returns_405_error(self):
+        """
+        Tests if request to Concent with will return HTTP 405 error
+        if not allowed HTTP method by view is used.
+        """
+
+        @api_view
+        @require_POST
+        def dummy_view(_request, _message):
+            assert False
+
+        request = self.request_factory.get(
+            "/dummy-url/",
+            content_type = 'application/octet-strea',
+            data         = '',
+        )
+        request.META['HTTP_CONCENT_CLIENT_PUBLIC_KEY'] = b64encode(settings.CONCENT_PUBLIC_KEY).decode('ascii')
+
+        response = dummy_view(request)  # pylint: disable=no-value-for-parameter,assignment-from-no-return
+
+        self.assertEqual(response.status_code,  405)
+
+
+def _log_message_received_500_mock(_message, _client_public_key):
+    Client.objects.get_or_create_full_clean(
+        CONCENT_PUBLIC_KEY
+    )
+    raise TypeError
+
+
+def _log_message_received_400_mock(_message, _client_public_key):
+    Client.objects.get_or_create_full_clean(
+        CONCENT_PUBLIC_KEY
+    )
+    raise Http400
+
+
+def gatekeeper_access_denied_response_500_mock(_message, _path = None, _subtask_id = None, _client_key = None):
+    Client.objects.get_or_create_full_clean(
+        CONCENT_PUBLIC_KEY
+    )
+    raise TypeError
+
+
+def gatekeeper_access_denied_response_400_mock(_message, _path = None, _subtask_id = None, _client_key = None):
+    Client.objects.get_or_create_full_clean(
+        CONCENT_PUBLIC_KEY
+    )
+    raise Http400
+
+
+@override_settings(
+    CONCENT_PRIVATE_KEY = CONCENT_PRIVATE_KEY,
+    CONCENT_PUBLIC_KEY  = CONCENT_PUBLIC_KEY,
+)
+class ApiViewTransactionTestCase(TransactionTestCase):
+
+    def test_api_view_should_rollback_changes_on_500_error(self):
+
+        with mock.patch('utils.logging.log_message_received', _log_message_received_500_mock):
+            try:
+                self.client.post(
+                    reverse('core:send'),
+                    data                                = '',
+                    content_type                        = 'application/octet-stream',
+                    HTTP_CONCENT_CLIENT_PUBLIC_KEY      = b64encode(settings.CONCENT_PUBLIC_KEY).decode('ascii'),
+                )
+            except TypeError:
+                pass
+
+        self.assertEqual(Client.objects.count(), 0)
```

Comment:
I think we need tests for the HTTP 200 cases when there is no rollback too.

Using `patch()` is tricky because you have to specify the module where the function is being used, not where it's defined. It's easy to make a mistake and break it silently. If this happens, we won't notice because the code still won't be creating any objects and the tests will pass. We will detect this only if we also have a test for HTTP 200 that uses the same mock - this one will break if the object is not created when it should.

Reply:
:+1: